cou/apps/base.py

```diff
@@ -569,28 +569,31 @@ def _get_upgrade_current_release_packages_step(
 
         return step
 
-    def _get_refresh_charm_step(self, target: OpenStackRelease) -> PreUpgradeStep:
-        """Get step for refreshing the charm.
+    def _get_refresh_charm_steps(self, target: OpenStackRelease) -> list[UpgradeStep]:
+        """Get steps for refreshing the charm.
 
         :param target: OpenStack release as target to upgrade
         :type target: OpenStackRelease
         :raises ApplicationError: When application has unexpected channel.
-        :return: Step for refreshing the charm
-        :rtype: PreUpgradeStep
+        :return: Steps for refreshing the charm
+        :rtype: list[PreUpgradeStep]
         """
         if self.is_from_charm_store:
-            return self._get_charmhub_migration_step(target)
+            return [self._get_charmhub_migration_step(target), self._get_wait_step()]
```

Comment:
I think this is a nice opportunity to create a `add_wait_step` decorator instead of changing individual functions. Something like:

```python
# add wait step in the last position of the list of steps
@add_wait_step(last=True)
def _get_refresh_charm_steps():
    pass
 
# add wait step in the first position of the list of steps
@add_wait_step(first=True)
def  post_upgrade_steps():
    pass
```

Reply:
I don't think so, because that `add_wait_step` needs context on the class (access to self.model, etc.), and it will depend on what upgrade steps are used.

I'm also very hesitant to refactor code here, especially refactoring while adding new features, because of the large effort to update the unit tests.  Already as it is, this is a 500 line change. :(